feat: update tokens-and-nfts/nfts-with-metaplex to use mpl-token-metadata (#339)
solana-foundation/developer-content: content/courses/tokens-and-nfts/nfts-with-metaplex.md

  files, any special traits the NFT has, and more.
- The **Metaplex Token Metadata** program is an onchain program that attaches
  metadata to a token mint. We can interact with the Token Metadata program
  using UMI and the

>   [Token Metadata plugin](https://developers.metaplex.com/token-metadata).

Reviewer: Can you change this PR to focus on `mpl-token-metadata` rather than `Umi`.  People searching for Umi get some [very complicated docs that won't help them make tokens](https://developers.metaplex.com/umi) but if we tell them they're using the `mpl-token-metadata` program they'll get the [better docs yu mentioned](https://developers.metaplex.com/token-metadata). 

Also Umi is in lowercase.

>   [Token Metadata plugin](https://developers.metaplex.com/token-metadata).

Reviewer: Can you change this PR to focus on `mpl-token-metadata` rather than `Umi`.  People searching for Umi get some [very complicated docs that won't help them make tokens](https://developers.metaplex.com/umi) but if we tell them they're using the `mpl-token-metadata` program they'll get the [better docs yu mentioned](https://developers.metaplex.com/token-metadata). 

Also Umi is in lowercase.

  files, any special traits the NFT has, and more.
- The **Metaplex Token Metadata** program is an onchain program that attaches
  metadata to a token mint. We can interact with the Token Metadata program

>   using Umi and the
>   [Token Metadata package](https://developers.metaplex.com/token-metadata) via
>   Umi, a tool made by Metaplex for working with onchain programs.
> 

Reviewer: ```suggestion
  using the[Token Metadata package](https://developers.metaplex.com/token-metadata) 
  via Umi, a tool made by Metaplex for working with onchain programs.

```

  offchain component of NFT metadata.

In the following sections, we'll cover the basics of using the

> `metaplex-foundation/token-metadata` plugin with UMI to prepare assets, create

Reviewer: Umi is in titlecase. Please do a find/replace.

NFTs, update NFTs, and associate an NFT with a broader collection. For more
information on `metaplex-foundation/token-metadata` see the
[developer docs for Token Metadata](https://developers.metaplex.com/token-metadata).

<Callout type="note">The Metaplex Foundation released a new standard for

> creating and managing NFTs called
> [Metaplex Core](https://developers.metaplex.com/core), that utilizes a single
> account model. But the only way to add metadata to SPL tokens is still done by
> using the Metadata program. </Callout>

Reviewer: ```suggestion
creating and managing NFTs called
[Metaplex Core](https://developers.metaplex.com/core), where an NFT simply has an owner, with the owner changing when the NFT is transferred. However the most common style of NFT is still by making a Solana SPL token with some Metadata attached via the Metaplex Metadata program, so that's what we'll be using in this tutorial.</Callout>
```

Can you simplify this section? Trying to avoid use of vague terms like 'account model' and mention specific differences. This is one suggestion .

#### UMI instance

> The UMI framework was designed by Metaplex to be a modular framework with zero
> dependencies that allow a developer to choose specific implementations/plugins
> that he will make use of and leave out those that won't be needed.
> 
> The framework also provides a set of default implementations and bundles out of
> the box that allow a developer to get started quickly.

Reviewer: Use simpler language.

```suggestion
Umi is a framework for making JS/TS clients for onchain program created by Metaplex. Umi can create JS/TS clients to many different programs, but in practice, it's most commonly used to communicate to the Token Metadata program.

Note Umi has different concepts for many concepts than web3.js, including Keypairs, PublicKeys, and Connections. However it is easy to convery from web3.j versions of these items to the Umi versions.
```

#### Installation and setting up Umi

First we create a new Umi instance. We can do this by either providing our own
RPC endpoint, or use the public facing Solana endpoints provided by the
`clusterApiUrl` method.

```typescript
import { createUmi } from "@metaplex-foundation/umi-bundle-defaults";
import { clusterApiUrl } from "@solana/web3.js";

const umi = createUmi(clusterApiUrl("devnet"));
```

> Finally, we pass in the identity/signer for our umi instance and the plugins

Reviewer: ```suggestion
Finally, we pass in the identity for our Umi instance (this is the keypair that will be used to sign transactions) and the plugins
```

> that we will make use of, in our case, this is the

Reviewer: ```suggestion
that we will use, in our case, this is the
```

`metaplex-foundation/mpl-token-metadata`.

```typescript

> import { mplTokenMetadata } from "@metaplex-foundation/mpl-token-metadata"; // [!code ++]

Reviewer: ```suggestion
import { mplTokenMetadata } from "@metaplex-foundation/mpl-token-metadata"; 
```

What was `// [!code ++]`?

import { keypairIdentity } from "@metaplex-foundation/umi"; // [!code ++]
import { createUmi } from "@metaplex-foundation/umi-bundle-defaults";
import { getKeypairFromFile } from "@solana-developers/helpers"; // [!code ++]

> import { clusterApiUrl } from "@solana/web3.js";

Reviewer: ```suggestion
import { clusterApiUrl } from "@solana/web3.js";
import { readFile } from "fs/promises";
```

const umi = createUmi(clusterApiUrl("devnet"));

// load keypair from local file system
// assumes that the keypair is already generated using `solana-keygen new`
const localKeypair = await getKeypairFromFile();

> // convert to umi compatible keypair

Reviewer: ```suggestion
// convert to Umi compatible keypair
```

const umiKeypair = umi.eddsa.createKeypairFromSecretKey(localKeypair.secretKey);

// load the MPL metadata program plugin and assign a signer to our umi instance
umi.use(keypairIdentity(umiKeypair)).use(mplTokenMetadata());
```

#### Uploading assets

Before creating an NFT, you must prepare and upload any assets you plan to
associate with the NFT. While this doesn't have to be an image, most NFTs have
an image associated with them.

Preparing and uploading an image involves converting the image to a buffer,
converting the file to a
[generic file](https://developers.metaplex.com/umi/storage#generic-files) using
the `createGenericFile()` function and finally uploading it to the designated
Storage Driver.

The `GenericFile` type allows Umi to support different file variations despite
the difference of browser files and local file system files i.e. those on your
computer.

In action, uploading an image names `random-image.png` from your computer would

> take the following steps,

Reviewer: ```suggestion
take the following steps:
```

> 1. Reading the file using `readFileSync` into a buffer.

Reviewer: ```suggestion
1. Reading the file using `readFile()` into a buffer.
```

We generally use promises where we can.

2. Creating a generic file type with the files MIME Type from the buffer and
   filePath.

3. Uploading file to designated storage provider.

```typescript

> let filePath = "random-image.png";
> 
> const buffer = readFileSync(filePath);

Reviewer: ```suggestion
let filePath = "random-image.png";

const buffer = await readFile(filePath);
```

let file = createGenericFile(buffer, filePath, {

>   contentType: "image/jpeg", // chose the correct file MIME type https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

Reviewer: ```suggestion
  // choose the correct file MIME type https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
  contentType: "image/jpeg", 
```

  offchain component of NFT metadata.

In the following sections, we'll cover the basics of using the
`metaplex-foundation/token-metadata` plugin with Umi to prepare assets, create
NFTs, update NFTs, and associate an NFT with a broader collection. For more
information on `metaplex-foundation/token-metadata` see the
[developer docs for Token Metadata](https://developers.metaplex.com/token-metadata).

<Callout type="note">
[Metaplex Core](https://developers.metaplex.com/core), is an NFT standard from Metaplex where asset details such as the owner, name, uri e.t.c are stored on a single account. However, the most common style of NFT is still by making a Solana
SPL token with some Metadata attached via the Metaplex Metadata program, so
that's what we'll be using in this tutorial. </Callout>

#### UMI instance

> Umi is a framework for making JS/TS clients for onchain program created by
> Metaplex. Umi can create JS/TS clients to many different programs, but in
> practice, it's most commonly used to communicate to the Token Metadata program.

Reviewer: ```suggestion
Umi is a framework for making JS/TS clients for onchain programs, that was created 
by Metaplex. Umi can create JS/TS clients for many different programs, but in
practice, it's most commonly used to communicate to the Token Metadata program.
```

[NFT standard](https://developers.metaplex.com/token-metadata/token-standard#the-non-fungible-standard)
to ensure compatibility with wallets, programs, and applications.

To create the metadata, use the `uploadJson()` method provided by the SDK. This
method accepts a metadata object and returns a URI that points to the uploaded
metadata.

```typescript
const uri = await umi.uploader.uploadJson({
  name,
  description,
  image,
});
```

#### Create the NFT

After uploading the NFT's metadata, you can finally create the NFT on the
network. The `mplTokenMetadata` plugin we added earlier provides the required
helpers to create an NFT or any other token with minimal configuration. The
helper `createNft` method will create the mint account, token account, metadata
account, and master edition account for you. The data provided to this method
will represent the onchain portion of the NFT metadata. You can explore the SDK
to see all the other input optionally supplied to this method.

```typescript
const { signature, result } = await createNft(umi, {
  mint,
  name: "My NFT",
  uri,
  updateAuthority: umi.identity.publicKey,
  sellerFeeBasisPoints: percentAmount(0),

> }).sendAndConfirm(umi, { send: { commitment: "finalized" } });

Reviewer: Question - does Umi allow us to set a default confirmation status?

Author: for the entire instance I am unsure.

Let me ask in the Metaplex discord.

Author: I'm guessing that if we instantiate it with the `Connection` object it inherits the commitment status 

```typescript
// create a new connection to Solana's devnet cluster
const connection = new Connection(clusterApiUrl("devnet"), {
	commitment: "finalized",
});
```

Author: The mentioned method wouldn't work -> [Metaplex Discord message link](https://discord.com/channels/848060988636921856/849844496941842432/1277874229928263764)

Solution provided is to create a wrapper function but that's too much to do -> [ Metaplex discord message link](https://discord.com/channels/848060988636921856/849844496941842432/1277965547773886576)

Reviewer: Thank Jimmi, I appreciate you chasing this with Metaplex. Understood.

provided.

This method returns an object containing the transaction signature and a result.
The result object contains the outcome of our transaction. If successful, the
`err` inside this will be set to null otherwise it'll contain the error for the
failed transaction.

By default, the SDK sets the `isMutable` property to true, allowing for updates
to be made to the NFT's metadata. However, you can choose to set `isMutable` to
false, making the NFT's metadata immutable.

#### Update the NFT

If you've left `isMutable` as true, you may update your NFT's metadata.

The SDK's `updateV1` method allows you to update both the onchain and offchain
portions of the NFT's metadata. To update the offchain metadata, you'll need to
repeat the steps of uploading a new image and metadata URI (as outlined in the
previous steps), then provide the new metadata URI to this method. This will
change the URI that the onchain metadata points to, effectively updating the
offchain metadata as well.

```typescript
const nft = await fetchMetadataFromSeeds(umi, { mintAddress });

await updateV1(umi, {
  mint,
  authority: umi.identity,
  data: {
    ...nft,

>     sellerFeeBasisPoints: 100, // 1%

Reviewer: I don't think it's worth setting `sellerFeeBasisPoints` - they're not enforcable, at least one marketplace decided to sell NFTs cheaper rather than pay artist royalties. Transfer hooks or pNFTs are a better solution for that problem.

Author: Umi has it configured as a non-nullable field. We have to include it.

Reviewer: Sorry , I should clarify: just set it to zero.

The last thing you need to do is verify the NFT. This effectively just flips the
`verified` field above to true, but it's incredibly important. This is what lets

> consuming programs and apps know that your NFT is in fact part of the

Reviewer: ```suggestion
consuming programs and apps, including wallets and art marketplaces, know that your NFT is in fact part of the
```

> collection. You can do this using the `verifyCollectionV1` function:

Reviewer: ```suggestion
collection - because the Collection's owner has signed a transaction making the NFT a member of that collection. You can do this using the `verifyCollectionV1` function:
```

```typescript
const metadata = findMetadataPda(umi, { mint: mint.publicKey });

await verifyCollectionV1(umi, {
  metadata,
  collectionMint,
  authority: umi.identity,
}).sendAndConfirm(umi);
```

### Lab

In this lab, we'll go through the steps to create an NFT using the Metaplex UMI
framework, update the NFT's metadata after the fact, and then associate the NFT
with a collection. By the end, you will have a basic understanding of how to use
the Metaplex UMI and the mplTokenMetadata library to interact with NFTs on
Solana.

#### Part 1: Creating an NFT collection

To begin, make a new folder and install the relevant dependencies:

```bash
npm i @solana/web3.js \
npm i @solana/web3.js \
npm i @solana-developers/helpers \
npm i @metaplex-foundation/mpl-token-metadata \
npm i @metaplex-foundation/umi-bundle-defaults \
npm i @metaplex-foundation/umi-uploader-irys \

> && npm i -save-dev esrun

Reviewer: Just do this like

```npm i 1 /
  2 /
  3 / 
  4 
```

To be consistent with our other content.

Also think it's `--save-dev`

```

Then create a file called `create-metaplex-collection.ts`, and add our imports:

```typescript
import {
  createNft,
  mplTokenMetadata,
} from "@metaplex-foundation/mpl-token-metadata";
import {
  createGenericFile,
  generateSigner,
  keypairIdentity,
  percentAmount,
} from "@metaplex-foundation/umi";
import { createUmi } from "@metaplex-foundation/umi-bundle-defaults";
import { irysUploader } from "@metaplex-foundation/umi-uploader-irys";
import {
  airdropIfRequired,
  getExplorerLink,
  getKeypairFromFile,
} from "@solana-developers/helpers";
import { clusterApiUrl, Connection, LAMPORTS_PER_SOL } from "@solana/web3.js";

> import { readFileSync } from "fs";

Reviewer: ```suggestion
import { readFile } from "fs/promises";
```

import * as path from "path";
```

Connect to devnet, load a user and Airdrop some SOL if needed:

```typescript
// create a new connection to Solana's devnet cluster
const connection = new Connection(clusterApiUrl("devnet"));

// load keypair from local file system

> // assumes that the keypair is already generated using `solana-keygen new`

Reviewer: Apply same comment as before

const collectionImagePath = path.resolve(__dirname, "collection.png");

const buffer = readFileSync(collectionImagePath);
let file = createGenericFile(buffer, collectionImagePath, {
  contentType: "image/png",
});
const [image] = await umi.uploader.upload([file]);
console.log("image uri:", image);

// upload offchain json to Arweave using irys
const uri = await umi.uploader.uploadJson({
  name: "My Collection",
  symbol: "MC",
  description: "My Collection description",
  image,
});
console.log("Collection offchain metadata URI:", uri);
```

Then actually make the collection:

```typescript
// generate mint keypair
const collectionMint = generateSigner(umi);

// create and mint NFT
await createNft(umi, {
  mint: collectionMint,
  name: "My Collection",
  uri,
  updateAuthority: umi.identity.publicKey,
  sellerFeeBasisPoints: percentAmount(0),
  isCollection: true,
}).sendAndConfirm(umi, { send: { commitment: "finalized" } });

let explorerLink = getExplorerLink(
  "address",
  collectionMint.publicKey,
  "devnet",
);
console.log(`Collection NFT:  ${explorerLink}`);
console.log(`Collection NFT address is:`, collectionMint.publicKey);
console.log("✅ Finished successfully!");
```

> We advise using [esrun]() to run the scripts because it allows you to use top

Reviewer: Link is missing.

```
% npx esrun create-metaplex-nft-collection.ts

Loaded user: 4kg8oh3jdNtn7j2wcS7TrUua31AgbLzDVkBZgTAe44aF
image uri: https://arweave.net/XWpt7HDOFC0wJQcQWgP9n_cxHS0qQik9-27CAAaGP6E
Collection offchain metadata URI: https://arweave.net/atIf58t3FHa3heoOtNqPkVvEGC_9WzAduY0GQE-LnFI
Collection NFT:  https://explorer.solana.com/address/D2zi1QQmtZR5fk7wpA1Fmf6hTY2xy8xVMyNgfq6LsKy1?cluster=devnet
Collection NFT address is: D2zi1QQmtZR5fk7wpA1Fmf6hTY2xy8xVMyNgfq6LsKy1
✅ Finished successfully!
```

Congratulations! You've created a Metaplex Collection. Check this out on Solana
Explorer using the URL above.

If you have any trouble, try and fix it yourself, but if you need to you can
also check out the

> [solution code](https://github.com/solana-developers/professional-education/blob/main/labs/metaplex-umi/create-collection.ts).

Reviewer: Add an image here of what Explorer should look like. See CONTRIBUTING for image specifications.

```
% npx esrun create-metaplex-nft.ts

Loaded user: 4kg8oh3jdNtn7j2wcS7TrUua31AgbLzDVkBZgTAe44aF
image uri: https://arweave.net/nVjCW_g-UMQUZano8vh-xPLX4nHwd5LsC8hXEuMdk_w
NFT offchain metadata URI: https://arweave.net/N2HivyLiwRdb4CutMSM9aDXMSMyS0PF_VSuAfKGUhhI
Token Mint:  https://explorer.solana.com/address/Zxd9TmtBHQNti6tJxtx1AKYJFykNUwJL4rth441CjRd?cluster=devnet
Created NFT address is Zxd9TmtBHQNti6tJxtx1AKYJFykNUwJL4rth441CjRd
✅ Finished successfully!
```

Inspect your NFT at the address given! If you have any trouble, try and fix it
yourself, but if you need to you can also check out the

> [solution code](https://github.com/solana-developers/professional-education/blob/main/labs/metaplex-umi/create-nft.ts).

Reviewer: Add an image here of what Explorer should look like. See CONTRIBUTING for image specifications.

await updateV1(umi, {
  mint,
  authority: umi.identity,
  data: {
    ...nft,
    sellerFeeBasisPoints: 100, // 1%
    name: "Updated Asset",
  },
  primarySaleHappened: true,
  isMutable: true,
}).sendAndConfirm(umi);

let explorerLink = getExplorerLink("address", mint, "devnet");
console.log(`NFT updated with new metadata URI: ${explorerLink}`);

console.log("✅ Finished successfully!");
```

Run `npx esrun update-metaplex-nft.ts`. You should see something like:

```bash
% npx esrun update-metaplex-nft.ts

Loaded user: 4kg8oh3jdNtn7j2wcS7TrUua31AgbLzDVkBZgTAe44aF
image uri: https://arweave.net/dboiAebucLGhprtknDQnp-yMj348cpJF4aQul406odg
NFT offchain metadata URI: https://arweave.net/XEjo-44GHRFNOEtPUdDsQlW5z1Gtpk2Wv0HvR8ll1Bw
NFT updated with new metadata URI: https://explorer.solana.com/address/Zxd9TmtBHQNti6tJxtx1AKYJFykNUwJL4rth441CjRd?cluster=devnet
✅ Finished successfully!
```

Inspect the updated NFT on Solana Explorer! Just like previously, if you have
any issues, you should fix them yourself, but if needed the
[solution code](https://github.com/solana-developers/professional-education/blob/main/labs/metaplex-umi/update-nft.ts)

> is available.

Reviewer: Add an image here of what Explorer should look like. See CONTRIBUTING for image specifications.
